src/data_structures/fmindex.rs

```diff
@@ -106,44 +86,102 @@ impl FMIndex {
         }
     }
 
-    pub fn occ(&self, r: usize, a: u8) -> usize {
+
+}
+
+/// The Fast Index in Minute space (FM-Index, Ferragina and Manzini, 2000) for finding suffix array
+/// intervals matching a given pattern.
+
+#[cfg_attr(feature = "serde_macros", derive(Serialize, Deserialize))]
+pub struct FMIndex<
+        DBWT: DerefBWT + Clone,
+        DLess: DerefLess + Clone,
+        DOcc: DerefOcc + Clone> {
+    bwt: DBWT,
+    less: DLess,
+    occ: DOcc,
+}
+
+impl<
+    DBWT: DerefBWT + Clone,
+    DLess: DerefLess + Clone,
+    DOcc: DerefOcc + Clone> FMIndexable for FMIndex<DBWT, DLess, DOcc> {
+
+    fn occ(&self, r: usize, a: u8) -> usize {
         self.occ.get(&self.bwt, r, a)
     }
-
-    pub fn less(&self, a: u8) -> usize {
+    fn less(&self, a: u8) -> usize {
         self.less[a as usize]
     }
-
     /// Provide a reference to the underlying BWT.
-    pub fn bwt(&self) -> &BWT {
+    fn bwt(&self) -> &BWT {
         &self.bwt
     }
 }
 
+impl<
+    DBWT: DerefBWT + Clone,
+    DLess: DerefLess + Clone,
+    DOcc: DerefOcc + Clone> FMIndex<DBWT, DLess, DOcc> {
+
+    /// Construct a new instance of the FM index.
+    ///
+    /// # Arguments
+    ///
+    /// * `sa` - the suffix array (or sample)
+    /// * `bwt` - the BWT
+    /// * `k` - the sampling rate of the occ array: every k-th entry will be stored (higher k means
+    ///   less memory usage, but worse performance)
+    /// * `alphabet` - the alphabet of the underlying text, omitting the sentinel
+    pub fn new(bwt: DBWT, less: DLess, occ: DOcc) -> Self {
+        FMIndex {
+            bwt: bwt,
+            less: less,
+            occ: occ,
+        }
+    }
+
+    /// Construct a new instance of the FMD index (see Heng Li (2012) Bioinformatics).
+    /// This expects a BWT that was created from a text over the DNA alphabet with N
+    /// (`alphabets::dna::n_alphabet()`) consisting of the
+    /// concatenation with its reverse complement, separated by the sentinel symbol `$`.
+    /// I.e., let T be the original text and R be its reverse complement.
+    /// Then, the expected text is T$R$. Further, multiple concatenated texts are allowed, e.g.
+    /// T1$R1$T2$R2$T3$R3$.
+    ///
+    pub fn as_fmdindex(self) -> FMDIndex<DBWT, DLess, DOcc> {
```

Comment:
Well, actually, that's only true if we're also okay with compromising on having it implement the FMIndexable trait. If we don't do that, we lose some ergonomics (`fmdindex.fmindex.smems()` instead of `fmdindex.smems()`) and we might as well get rid of `FMIndexable`. Would you prefer that? I'll make the `From` impl right away, but I'm not as sure about the other change.